src/main/java/sbt_inc/SbtIncrementalCompiler.java

```diff
@@ -111,10 +112,14 @@ public SbtIncrementalCompiler(
                 }
                 : new URL[] {libraryJar.toURI().toURL(), compilerJar.toURI().toURL()};
 
+    URLClassLoader classLoader =
+        scalaVersion.major == 3
+            ? new URLClassLoader(urls, SbtIncrementalCompiler.class.getClassLoader())
```

Comment:
I am not 100% sure but I think the `SbtIncrementalCompiler.class.getClassLoader()` contains the Scala 2 compiler jars. Those will hide the Scala 3 compiler classes and prevent us from compiling in Scala 3.

What solution do we have?

## 1. Use a Layered class loader

Invoke `SbtIncrementalCompiler` in a layered class loader.

![](https://mermaid.ink/img/eyJjb2RlIjoiZ3JhcGggVERcbiAgICBcbiAgICBBW3NjYWxhLW1hdmVuLXBsdWdpbi5qYXI8YnIvPnNjYWxhLWNvbXBpbGVyLmphcjxici8-c2NhbGEtbGlicmFyeS5qYXI8YnIvPi4uLl0gLS0-fHBhcmVudHwgQlxuICAgIEJbY29tcGlsZXItaW50ZXJmYWNlLmphcl0gLS0-fHBhcmVudHxDXG4gICAgQ1tudWxsXVxuICAgIERbc2NhbGEzLWNvbXBpbGVyLmphcjxici8-c2NhbGEzLWxpYnJhcnkuamFyPGJyLz4uLi5dIC0tPiB8cGFyZW50fEIiLCJtZXJtYWlkIjp7fSwidXBkYXRlRWRpdG9yIjpmYWxzZX0)

To build the Scala 3 compiler class loader you can get the `compiler-interface.jar` class loader:

```scala
? new URLClassLoader(urls, xsbti.Reporter.class.getClassLoader(xsbti.Reporter.class.getClassLoader()))
```

## 2. Create a filtered class loader

Encapsulate `SbtIncrementalCompiler.class.getClassLoader()` in a custom `Scala3CompilerLoader` that resolves the `xsbti.*` classes (classes from `compiler-interface.jar`)  from the `SbtIncrementalCompiler.class.getClassLoader()`.

```java
? new Scala3CompilerLoader(urls, xsbti..class.getClassLoader(xsbti.Reporter.class.getClassLoader()))
```
Where:

```java
class Scala3CompilerLoader extends URLClassLoader {
  private final ClassLoader sbtLoader;

  public CompilerClassLoader(URL[] urls, ClassLoader sbtLoader) {
    super(urls, null);
    this.sbtLoader = sbtLoader;
  }

  @Override
  public Class<?> loadClass(String className, boolean resolve) throws ClassNotFoundException {
    if (className.startsWith("xsbti.")) {
      // We can't use the loadClass overload with two arguments because it's
      // protected, but we can do the same by hand (the classloader instance
      // from which we call resolveClass does not matter).
      Class<?> c = sbtLoader.loadClass(className);
      if (resolve)
        resolveClass(c);
      return c;
    } else {
      return super.loadClass(className, resolve);
    }
  }
}
```

Reply:
@adpi2 `new URLClassLoader(urls, xsbti.Reporter.class.getClassLoader(xsbti.Reporter.class.getClassLoader()))` is not compiling. you meant `new URLClassLoader(urls, xsbti.Reporter.class.getClassLoader())`?

Reply:
@adpi2 `new Scala3CompilerLoader(urls, xsbti..class.getClassLoader(xsbti.Reporter.class.getClassLoader()))` is also gives compile error. could you please take a look at my latest commit?

Reply:
Your latest commit looks good